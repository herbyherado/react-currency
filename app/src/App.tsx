import React, { Component } from 'react';
import { observer } from 'mobx-react';

import CurrencyInput from './components/CurrencyInput';
import CurrencyCard from './components/CurrencyCard';
import CurrencyAdd from './components/CurrencyAdd';

import store, { CurrencyValues, CurrencyNames } from './stores/currency';
import { DropdownOption } from './components/CurrencyAdd';
import './App.css';

@observer
class App extends Component<AppProps, AppState> {
  state: AppState = {
    baseValue: 10,
    _addNew: false
  };

  handleInputChange = (e: React.FormEvent<HTMLInputElement>) => {
    const bigint = 9007199254740992; // maximum number allowed in JS 2^53
    let input = e.currentTarget.value;
    let selectedValue = Number(input) > bigint ? input : Number(input);
    if (!isNaN(Number(input))) {
      this.setState({ baseValue: selectedValue });
    }
  };

  handleRemoveCurrency = (index: number) => {
    if (store.countDisplayed > 1) {
      store.removeCurrency(index);
    }
  };

  handleAddCurrency = () => this.setState({ _addNew: true });

  handleSelectCurrency = (option: DropdownOption) => {
    store.addCurrency(option.value);
    this.setState({ _addNew: false });
  };

  render() {
    const { baseValue, _addNew } = this.state;
    const { baseCurrency, displayCurrencies, currencyValues, names } = this.props.store;

    return (
      <div className="App">
        <div className="App-header">
          <div className="currency-table">
            <CurrencyInput
              baseName={baseCurrency}
              baseFullName={store.getCurrencyName(baseCurrency)}
              baseValue={baseValue}
              onInputChange={(e: React.FormEvent<HTMLInputElement>) => this.handleInputChange(e)}
            />
            <div className="currency-content">
              {displayCurrencies.map((currency: string, index: number) => (
                <CurrencyCard
                  key={`currency-${index}`}
                  currencyName={currency}
                  currencyFullName={names[currency]}
                  currencyValue={currencyValues[currency]}
                  baseName={baseCurrency}
                  baseValue={baseValue}
                  onRemoveCard={() => this.handleRemoveCurrency(index)}
                />
              ))}
              <CurrencyAdd
                addNew={_addNew}
                options={store.currencyList}
                onSelectCurrency={this.handleSelectCurrency}
                onAddCurrency={this.handleAddCurrency}
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

type AppState = {
  baseValue: number | string;
  _addNew: boolean;
};

type AppProps = {
  store: {
    baseCurrency: string;
    displayCurrencies: string[];
    currencyValues: CurrencyValues;
    names: CurrencyNames;
  };
};

export default App;
