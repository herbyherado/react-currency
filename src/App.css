* {
  box-sizing: border-box;
}

p {
  margin: 0;
}

.text-italic {
  font-weight: 500;
  font-style: italic;
}

.text-bolder {
  font-weight: 700;
}

.App {
  font-size: 14px;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;

  padding: 4rem 0 16rem 0;
  color: white;
}

/* CONTENT */

div.currency-content {
  padding: 0 1rem;
  min-height: 540px;
}
