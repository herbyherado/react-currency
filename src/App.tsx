import React, { Component } from 'react';
import { observer } from 'mobx-react';

import CurrencyInput from './components/CurrencyInput';
import CurrencyCard from './components/CurrencyCard';
import CurrencyAdd from './components/CurrencyAdd';

import currencyStore, { CurrencyValues, CurrencyNames } from './stores/currency';
import { DropdownOption } from './components/CurrencyAdd';
import './App.css';

@observer
class App extends Component<AppProps, AppState> {
  state: AppState = {
    baseValue: 10,
    _addNew: false
  };

  handleInputChange = (e: React.FormEvent<HTMLInputElement>) => {
    if (!isNaN(Number(e.currentTarget.value))) {
      this.setState({ baseValue: Number(e.currentTarget.value) });
    }
  };

  handleRemoveCurrency = (index: number) => {
    if (currencyStore.countDisplayed > 1) {
      currencyStore.removeCurrency(index);
    }
  };

  handleAddCurrency = () => this.setState({ _addNew: true });

  handleSelectCurrency = (option: DropdownOption) => {
    currencyStore.addCurrency(option.value);
    this.setState({ _addNew: false });
  };

  render() {
    const { baseValue, _addNew } = this.state;
    const { baseCurrency, displayCurrencies, currencyValues, names } = this.props.store;

    return (
      <div className="App">
        <div className="App-header">
          <div className="currency-table">
            <CurrencyInput
              baseName={baseCurrency}
              baseFullName={currencyStore.getCurrencyName(baseCurrency)}
              baseValue={baseValue}
              onInputChange={(e: React.FormEvent<HTMLInputElement>) => this.handleInputChange(e)}
            />
            <div className="currency-content">
              {displayCurrencies.map((currency: string, index: number) => (
                <CurrencyCard
                  key={`currency-${index}`}
                  currencyName={currency}
                  currencyFullName={names[currency]}
                  currencyValue={currencyValues[currency]}
                  baseName={baseCurrency}
                  baseValue={baseValue}
                  onRemoveCard={() => this.handleRemoveCurrency(index)}
                />
              ))}
              <CurrencyAdd
                addNew={_addNew}
                options={currencyStore.currencyList}
                onSelectCurrency={this.handleSelectCurrency}
                onAddCurrency={this.handleAddCurrency}
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

type AppState = {
  baseValue: number;
  _addNew: boolean;
};

type AppProps = {
  store: {
    baseCurrency: string;
    displayCurrencies: string[];
    currencyValues: CurrencyValues;
    names: CurrencyNames;
  };
};

export default App;
