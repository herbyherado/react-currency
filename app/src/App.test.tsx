import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from './App';

import store from './stores/currency';

it('renders without crashing', function() {
  const div = document.createElement('div');
  ReactDOM.render(<App store={store} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('App component testing', function() {
  describe('App snapshot test', function() {
    it('Should render to match snapshot', function() {
      const component = renderer.create(<App store={store} />);
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  describe('App method calls', function() {
    const component = renderer.create(<App store={store} />);

    describe('handleInputChange', function() {
      const setStateSpy = jest.spyOn(App.prototype, 'setState');
      const handleInputChange = component.root.instance.handleInputChange;

      it('should not call setState with incorrect input', function() {
        const mockedEvent = { currentTarget: { value: '  a12' } };
        handleInputChange(mockedEvent);
        expect(setStateSpy).toBeCalledTimes(0);
      });

      it('should call setState with correct input', function() {
        const mockedEvent = { currentTarget: { value: '12' } };

        handleInputChange(mockedEvent);
        expect(setStateSpy).toHaveBeenCalled();
        expect(setStateSpy).toBeCalledTimes(1);
      });
    });

    describe('handleRemoveCurrency', function() {
      test('should successfully call store action', function() {
        const handleRemoveCurrency = component.root.instance.handleRemoveCurrency;
        const displayCurrencies = store.displayCurrencies;
        const storeSpy = jest.spyOn(store, 'removeCurrency');
        expect(store.countDisplayed).toBe(4);

        handleRemoveCurrency(1);
        expect(displayCurrencies).toHaveLength(3);
        expect(storeSpy).toHaveBeenCalled();
      });

      test('should not call store action', function() {
        const handleRemoveCurrency = component.root.instance.handleRemoveCurrency;
        const displayCurrencies = store.displayCurrencies;

        expect(store.countDisplayed).toBe(3);
        handleRemoveCurrency(2);
        handleRemoveCurrency(1);
        expect(displayCurrencies).toHaveLength(1);
        handleRemoveCurrency(0);
        expect(displayCurrencies).toHaveLength(1);
      });
    });

    describe('handleAddCurrency', function() {
      const setStateSpy = jest.spyOn(App.prototype, 'setState');
      const handleAddCurrency = component.root.instance.handleAddCurrency;

      test('should call setState', function() {
        handleAddCurrency('CAD');
        expect(setStateSpy).toBeCalled();
      });
    });

    describe('handleSelectCurrency', function() {
      const setStateSpy = jest.spyOn(App.prototype, 'setState');
      const handleSelectCurrency = component.root.instance.handleSelectCurrency;

      test('should call state & action', function() {
        const mockOption = { value: 'CAD', label: 'Canadian Dollar' };
        handleSelectCurrency(mockOption);
        expect(store.displayCurrencies).toContain('CAD');
        expect(setStateSpy).toBeCalled();
      });
    });
  });
});
