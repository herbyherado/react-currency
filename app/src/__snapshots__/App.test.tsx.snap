// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component testing App snapshot test Should render to match snapshot 1`] = `
<div
  className="App"
>
  <div
    className="App-header"
  >
    <div
      className="currency-table"
    >
      <div
        className="currency-input"
      >
        <p
          className="currency-input-detail text-italic"
        >
          USD
           - 
          United States Dollar
        </p>
        <div
          className="currency-input-body"
        >
          <p
            className="currency-input-text"
          >
            USD
          </p>
          <input
            onChange={[Function]}
            placeholder="Value"
            type="text"
            value={10}
          />
        </div>
      </div>
      <div
        className="currency-content"
      >
        <div
          className="card"
        >
          <div
            className="card-content"
          >
            <div
              className="card-name"
            >
              <p>
                IDR
              </p>
            </div>
            <div
              className="card-footer"
            >
              <p
                className="text-italic text-bolder"
              >
                IDR
                 - 
                Indonesian Rupiah
              </p>
              <p
                className="text-italic"
              >
                1 
                USD
                 = 
                IDR
                 
              </p>
            </div>
          </div>
          <button
            className="card-delete"
            onClick={[Function]}
          >
            -
          </button>
        </div>
        <div
          className="card"
        >
          <div
            className="card-content"
          >
            <div
              className="card-name"
            >
              <p>
                EUR
              </p>
            </div>
            <div
              className="card-footer"
            >
              <p
                className="text-italic text-bolder"
              >
                EUR
                 - 
                Euro
              </p>
              <p
                className="text-italic"
              >
                1 
                USD
                 = 
                EUR
                 
              </p>
            </div>
          </div>
          <button
            className="card-delete"
            onClick={[Function]}
          >
            -
          </button>
        </div>
        <div
          className="card"
        >
          <div
            className="card-content"
          >
            <div
              className="card-name"
            >
              <p>
                GBP
              </p>
            </div>
            <div
              className="card-footer"
            >
              <p
                className="text-italic text-bolder"
              >
                GBP
                 - 
                British Pound Sterling
              </p>
              <p
                className="text-italic"
              >
                1 
                USD
                 = 
                GBP
                 
              </p>
            </div>
          </div>
          <button
            className="card-delete"
            onClick={[Function]}
          >
            -
          </button>
        </div>
        <div
          className="card"
        >
          <div
            className="card-content"
          >
            <div
              className="card-name"
            >
              <p>
                SGD
              </p>
            </div>
            <div
              className="card-footer"
            >
              <p
                className="text-italic text-bolder"
              >
                SGD
                 - 
                Singapore Dollar
              </p>
              <p
                className="text-italic"
              >
                1 
                USD
                 = 
                SGD
                 
              </p>
            </div>
          </div>
          <button
            className="card-delete"
            onClick={[Function]}
          >
            -
          </button>
        </div>
        <button
          className="currency-add"
          onClick={[Function]}
          type="button"
        >
          Add More Currencies
        </button>
      </div>
    </div>
  </div>
</div>
`;
